proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Accompanying file `headerparser.policy` contains an example policy
+// for this protocol. Install it with:
+// $ cilium policy import proxylib/testparsers/headerparser.policy
+//
+
+package binarymemcached
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/proxylib/proxylib"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// BinaryMemcacheRule matches against memcached requests
+type BinaryMemcacheRule struct {
+	// opCode group name
+	opCode string
+
+	keyExact  string
+	keyPrefix string
+	keyRegex  string
+
+	// allowed opcodes
+	opCodes []byte
+	// compiled regex
+	regex *regexp.Regexp
+}
+
+// Matches returns true if the BinaryMemcacheRule matches
+func (rule *BinaryMemcacheRule) Matches(data interface{}) bool {
+	log.Infof("binarymemcache checking rule %v", *rule)
+
+	packet := data.([]byte)
+	key, err := getMemcacheKey(packet)
```

Comment:
Here the data is parsed again for each rule match attempt. You could pass in a pre-parsed message as `data` instead. See comment in `OnData()` below.

Reply:
done